File: src/turbopuffer_fs/ingest.py
"""Directory ingest helpers."""

from __future__ import annotations

from pathlib import Path

from .live import mount_namespace
from .schema import directory_row, row_from_bytes, upsert_rows_payload


def mounted_path(local_root: Path, local_path: Path, mount_root: str = "/") -> str:
    from .paths import join_path, normalize_path

    root = normalize_path(mount_root)
    relative = local_path.relative_to(local_root).as_posix()
    return root if relative == "." else join_path(root, relative)


def scan_directory(local_root: str | Path, *, mount_root: str = "/") -> list[dict[str, object]]:
    root = Path(local_root).expanduser().resolve()
    if not root.exists():
        raise FileNotFoundError(str(root))
    if not root.is_dir():
        raise NotADirectoryError(str(root))

    rows: list[dict[str, object]] = []
    for local_path in [root, *sorted(root.rglob("*"))]:
        mount_path = mounted_path(root, local_path, mount_root=mount_root)
        stat = local_path.stat()
        if local_path.is_dir():
            rows.append(directory_row(mount_path, source_mtime_ns=stat.st_mtime_ns))
            continue
        rows.append(
            row_from_bytes(
                mount_path,
                local_path.read_bytes(),
                source_mtime_ns=stat.st_mtime_ns,
                source_size_bytes=stat.st_size,
            )
        )
    return rows


def batched(rows: list[dict[str, object]], batch_size: int) -> list[list[dict[str, object]]]:
    size = int(batch_size)
    if size < 1:
        raise ValueError("batch_size must be a positive integer")
    return [rows[index : index + size] for index in range(0, len(rows), size)]


def write_rows(client, namespace: str, rows: list[dict[str, object]], *, batch_size: int = 256) -> list[dict[str, object]]:
    handle = client.namespace(namespace)
    responses = []
    for batch in batched(rows, batch_size):
        responses.append(handle.write(**upsert_rows_payload(batch)))
    return responses


def ingest_directory(client, mount: str, local_root: str | Path, *, mount_root: str = "/", batch_size: int = 256) -> dict[str, object]:
    rows = scan_directory(local_root, mount_root=mount_root)
    writes = write_rows(client, mount_namespace(mount), rows, batch_size=batch_size)
    return {
        "mount": mount,
        "namespace": mount_namespace(mount),
        "row_count": len(rows),
        "rows": rows,
        "writes": writes,
    }

File: packages/turbopuffer-fs/src/ingest.ts
import { readdir, readFile, stat as fsStat } from "node:fs/promises";
import { resolve as resolvePath } from "node:path";

import { mountNamespace } from "./live.js";
import { joinPath, normalizePath } from "./paths.js";
import { directoryRow, rowFromBytes, upsertRowsPayload } from "./schema.js";
import type { AnyObject, FsRow } from "./types.js";

export function mountedPath(localRoot: string, localPath: string, mountRoot = "/"): string {
  const normalizedMountRoot = normalizePath(mountRoot);
  const relative = localPath.slice(localRoot.length).replace(/^\/+/, "");
  return relative === "" ? normalizedMountRoot : joinPath(normalizedMountRoot, relative);
}

async function scanRecursive(root: string, path: string, rows: FsRow[], mountRoot: string): Promise<void> {
  const stats = await fsStat(path);
  const mountPath = mountedPath(root, path, mountRoot);
  if (stats.isDirectory()) {
    rows.push(directoryRow(mountPath, Number(stats.mtimeMs * 1_000_000)));
    const entries = (await readdir(path)).sort();
    for (const entry of entries) {
      await scanRecursive(root, resolvePath(path, entry), rows, mountRoot);
    }
    return;
  }
  const data = new Uint8Array(await readFile(path));
  rows.push(
    rowFromBytes(
      mountPath,
      data,
      undefined,
      Number(stats.mtimeMs * 1_000_000),
      stats.size,
    ),
  );
}

export async function scanDirectory(localRoot: string, options: { mountRoot?: string } = {}): Promise<FsRow[]> {
  const root = resolvePath(localRoot);
  const stats = await fsStat(root);
  if (!stats.isDirectory()) {
    throw new Error(`NotADirectoryError:${root}`);
  }
  const rows: FsRow[] = [];
  await scanRecursive(root, root, rows, options.mountRoot ?? "/");
  return rows;
}

export function batched<T>(rows: T[], batchSize: number): T[][] {
  if (!Number.isInteger(batchSize) || batchSize < 1) {
    throw new Error("batchSize must be a positive integer");
  }
  const batches: T[][] = [];
  for (let index = 0; index < rows.length; index += batchSize) {
    batches.push(rows.slice(index, index + batchSize));
  }
  return batches;
}

export async function writeRows(
  client: { namespace(namespace: string): { write(payload: AnyObject): Promise<unknown> } },
  namespace: string,
  rows: FsRow[],
  options: { batchSize?: number } = {},
): Promise<unknown[]> {
  const handle = client.namespace(namespace);
  const batchSize = options.batchSize ?? 256;
  const responses: unknown[] = [];
  for (const batch of batched(rows, batchSize)) {
    responses.push(await handle.write(upsertRowsPayload(batch) as AnyObject));
  }
  return responses;
}

export async function ingestDirectory(
  client: { namespace(namespace: string): { write(payload: AnyObject): Promise<unknown> } },
  mount: string,
  localRoot: string,
  options: { mountRoot?: string; batchSize?: number } = {},
): Promise<AnyObject> {
  const rows = await scanDirectory(localRoot, { mountRoot: options.mountRoot });
  const namespace = mountNamespace(mount);
  const writes = await writeRows(client, namespace, rows, { batchSize: options.batchSize });
  return {
    mount,
    namespace,
    rowCount: rows.length,
    rows,
    writes,
  };
}
